[FIX] l10n_ar_tax: Cálculo de retención de ganancias.

pablohmontenegro opened the pull request:
Tarea: 87686
Cuando el contacto tiene el campo imp_ganancias_padron (Ganancias) distinto de "Activo" entonces no debe calcular retención de ganancias sin importar si la posición fiscal tiene retención de ganancias.

roboadhoc commented:
[Pull request status dashboard](https://mergebot.adhoc.ar/ingadhoc/odoo-argentina/pull/1049).

pablohmontenegro committed:
[FIX] l10n_ar_tax: Cálculo de retención de ganancias.
Tarea: 87686
Cuando el contacto tiene el campo imp_ganancias_padron (Ganancias) distinto de "Activo" entonces no debe calcular retención de ganancias sin importar si la posición fiscal tiene retención de ganancias.

zaoral commented on l10n_ar_tax/models/account_fiscal_position.py:
```diff
@@ -10,7 +11,23 @@ def _l10n_ar_add_taxes(self, partner, company, date, tax_type):
+                    self.env._(
+                        "El contacto %s no tiene configurada inscripcion en impuesto a las ganancias" % partner.name
+                    )
+                )
```
Do it before the for, In an if and use a filtered. 
you can show the detail info of the error with the problematic partners

zaoral commented on l10n_ar_tax/models/account_fiscal_position.py:
```diff
@@ -10,7 +11,28 @@ def _l10n_ar_add_taxes(self, partner, company, date, tax_type):
         self.ensure_one()
         taxes = self.env["account.tax"]
-        for fp_tax in self.l10n_ar_tax_ids.filtered(lambda x: x.tax_type == tax_type):
+        # Si se va a calcular retención de ganancias previamente queremos que tenga configurado el partner
+        # La inscripción imp_ganancias_padron que corresponda.
+        if self.l10n_ar_tax_ids.filtered(
+            lambda x: x.tax_type == tax_type
+            and x.default_tax_id.l10n_ar_tax_type in ["earnings", "earnings_scale"]
+            and partner.imp_ganancias_padron == False
+        ):
+            raise UserError(
+                self.env._("El contacto %s no tiene configurada inscripcion en impuesto a las ganancias" % partner.name)
+            )
+        # El filtered en la parte del "and" va a dar true para todo impuesto que no sea de ganancias.
+        # Si llega a ser de ganancias va a dar False en los casos que no queremmos que se calcule que son
+        # cuando no está inscripto o no corresponde. Si está activo en la inscripción de ganancias
+        # entonces va a dar True y calculará la retención de ganancias.
+        for fp_tax in self.l10n_ar_tax_ids.filtered(
+            lambda x: x.tax_type == tax_type
+            and (
+                x.default_tax_id.l10n_ar_tax_type not in ["earnings", "earnings_scale"]
+                if partner.imp_ganancias_padron != "AC"
+                else True
+            )
+        ):
```
```suggestion
        for fp_tax in self.l10n_ar_tax_ids.filtered( lambda x: x.tax_type == tax_type)
            if fp_tax.default_tax_id.l10n_ar_tax_type in ["earnings", "earnings_scale"]>
                 if  partner.imp_ganancias_padron == False
                      raise UserError(
                          self.env._("El contacto %s no tiene configurada inscripcion en impuesto a las ganancias" % partner.name)
                      )
               if partner.imp_ganancias_padron != "AC"
                   continue
        ):
```

jjscarafia commented:
cerramos porque al final vamos a ocultar los datos ya que no se usarían
